WooCommerce/Classes/ViewRelated/Products/Media/ProductImagesSaver.swift

```diff
@@ -0,0 +1,99 @@
+import Combine
+import Yosemite
+
+final class ProductImagesSaver {
+    /// Initially set when product save is requested when all images at the time of request are uploaded.
+    /// As each pending image is uploaded, the statuses are also updated and ready to save when none is pending upload.
+    private(set) var imageStatusesToSave: [ProductImageStatus] = [] {
+        didSet {
+            imageStatusesToSaveSubject.send(imageStatusesToSave)
+        }
+    }
+    // The use of a `PassthroughSubject` sent on variable `didSet` instead of a `@Published` is because `@Published`
+    // subscription happens in `willSet` while we want to update the statuses after `didSet`.
+    private let imageStatusesToSaveSubject: PassthroughSubject<[ProductImageStatus], Never> = .init()
+
+    private var uploadStatusesSubscription: AnyCancellable?
+    private var assetUploadSubscription: AnyCancellable?
+
+    private let siteID: Int64
+    private let productID: Int64
+    private let stores: StoresManager
+
+    init(siteID: Int64, productID: Int64, stores: StoresManager = ServiceLocator.stores) {
+        self.siteID = siteID
+        self.productID = productID
+        self.stores = stores
+    }
+
+    /// If any images are pending upload, saves the product remotely with the uploaded images when none is pending upload.
+    /// - Parameters:
+    ///   - imageActionHandler: action handler that provides the latest image statuses and image asset upload subscription.
+    ///   - onProductSave: called after the product is updated remotely with the uploaded images.
+    func saveProductImagesWhenNoneIsPendingUploadAnymore(imageActionHandler: ProductImageActionHandler,
+                                                         onProductSave: @escaping (Result<[ProductImage], Error>) -> Void) {
+        let imageStatuses = imageActionHandler.productImageStatuses
+        guard imageStatuses.hasPendingUpload else {
+            return
+        }
+
+        imageStatusesToSave = imageStatuses
+
+        uploadStatusesSubscription = imageStatusesToSaveSubject
+            .filter { $0.hasPendingUpload == false }
+            .map { $0.images }
+            .filter { $0.isNotEmpty }
+            .sink(receiveValue: { [weak self] images in
+                self?.saveProductImages(images, onProductSave: onProductSave)
+            })
+
+        assetUploadSubscription = imageActionHandler.addAssetUploadObserver(self) { [weak self] asset, result in
```

Comment:
Super nit: What do you think about moving the `uploadStatusesSubscription` and `assetUploadSubscription` creation logic into private methods to improve readability?